exercises/structs/structs3.rs

```diff
@@ -7,26 +7,26 @@
 
 #[derive(Debug)]
 struct Package {
-    from: String,
-    to: String,
-    weight: f32
+    sender_country: String,
+    recipient_country: String,
+    weight_in_grams: i32,
 }
 
 impl Package {
-    fn new(from: String, to: String, weight: f32) -> Package {
-        if weight <= 0.0 {
+    fn new(sender_country: String, recipient_country: String, weight_in_grams: i32) -> Package {
+        if weight_in_grams <= 0 {
             // Something goes here...
         } else {
-            return Package {from, to, weight};
+            return Package {sender_country, recipient_country, weight_in_grams};
         }
     }
 
     fn is_international(&self) -> ??? {
         // Something goes here...
     }
 
-    fn get_fees(&self, cost_per_kg: f32) -> ??? {
-        // Something goes here...
+    fn get_fees(&self, cents_per_kg: i32) -> ??? {
```

Comment:
Perhaps this should also be `cents_per_gram`. Otherwise, the output may have to be a float, which we're trying to avoid. For example, a package weighing 150 grams -- with a cost of 10 cents per kg -- would have a cost of 1.5 cents.

Reply:
1.5 cent will be rounded and that's what we want because we can't pay 0.5 cents anyway.
Cents per gram is not precise enough as a price is usually less than one cent per gram.
Yes it require to divide by 1000. I think that is not hard to grasp. I also think it would be nice to add a link to a good explanation why float are bad for money.

Reply:
Currently, it asks the user to supply the return type. Many users likely will reason that it should be a float, since otherwise information is lost (and as-is it is unclear whether they should round or truncate). 

So, to avoid another issue, it might be good to either:
1. Force the return type to be an integer, and add a comment asking the user to truncate the result. -or-
2. Use `cents_per_gram`, where we will not have to deal with fractions.

Reply:
There is no need to truncate or round, integers operations do it by themselves.
Relevant information is lost with floats, not with integer.
I think your option 1 is the good one, no need to ask for any truncate or rounding. It is done naturally by integers operation.
Option 2 makes no sense as even the smallest integer price for 1 gram is out of scale ($1000 for 1kg).

Reply:
Thinking through this as a student, I would notice that `1254 / 1000 == 1`. I would be hesitant to lose the extra 254 grams via an integer division.

Since Rustlings did not specify how to handle excess grams, I would likely decide the safest approach is to pro-rate `cents_per_kg`, maybe like this:

> let weight_in_kgs = (weight_in_grams as f32) / 1000.;
> return weight_in_kgs * (cents_per_kg as f32) as i32; 

Yet even now, I am unsure whether I wrote the function as Rustlings intended -- should I have rounded or truncated the result? (Money is often rounded.)

So, this is why I recommend including a comment explaining exactly how the fees should be computed. I am just trying to avoid another Issue later on 😀

Reply:
You complicate your life too much ! 😀 You must multiply first then divide.
`
fn get_fees(&self, cents_per_kg: i32) -> i32 {
    self.weight_in_grams * cents_per_kg / 1000
}
`

No need for cast or anything.

Reply:
We could also simply store weight in kg and integer, ignoring grams for the example it makes everything more simple and is more correct than using floats for money.

Reply:
😀 Yes, that is a much better method and yields the same result! So perhaps a specification would be something like:

``` rust
// Fees are pro-rated. Disregard any final fractions of a cent.
// Try to compute the fees without using floats!
```

Reply:
As you want as long there's no float used for money, I'm happy.

I just did the clippy1 exercice with the beautifull advice :
```
  if y != x {
   |        ^^^^^^ help: consider comparing them within some error: `(y - x).abs() > error`
```
That's how floats must be handled !

Reply:
> We could also simply store weight in kg and integer, ignoring grams for the example it makes everything more simple and is more correct than using floats for money. Should I do another pull request this way ?

It might be easiest just to make them all the same units. Other than this, I think it looks great!

Reply:
I'm a bit unsure about the units puzzle: it risks losing focus, but a bit of non-Rust-related challenge makes the exercises more amenable. No strong opinion anyway.